### PR TITLE
wgsl: Add validation tests for tokenization

### PR DESCRIPTION
**Pull request by jrprice**
Covers the following cases:
- Empty module
- Null characters
- Whitespace (delimiters)
- Comments
- Identifiers

Tested and mostly passing with ToT Dawn+Tint, with the following exceptions:
- Null characters are terminating the source string rather than being rejected
- The `enable` keyword is not yet reserved by Tint

<hr>

**Requirements for PR author:**

- [X] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [X] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [X] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/shader/validation/tokens.spec.ts [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+  )
+  .fn(t => {
+    let code = '';
+    if (t.params.placement === 'comment') {
+      code = `// Here is a ${t.params.contains_null ? '\0' : 'Z'} character`;
+    } else if (t.params.placement === 'delimiter') {
+      code = `let${t.params.contains_null ? '\0' : ' '}name : i32 = 0;`;
+    } else if (t.params.placement === 'eol') {
+      code = `let name : i32 = 0;${t.params.contains_null ? '\0' : ''}`;
+    }
```
Comment (jrprice) [inline src/webgpu/shader/validation/tokens.spec.ts:31]: @kainino0x or somebody else that understands Javascript: I'm not actually sure whether this is correct or not. Should these JS strings containing null characters get passed through the WebGPU implementation verbatim, or does something else cause them to be null terminated?

Comment (kainino0x) [reply src/webgpu/shader/validation/tokens.spec.ts:31]: Not 100% sure if this answers the question, but:

JS strings are not null-terminated, internally they're essentially a `vector<uint16_t>`.
When we get a string in Blink for example, it arrives as WTF::String or something like that, and we can convert to UTF-8, and NULs inside the string get carried through.
(NUL is the 0 codepoint, encoded as 0x0000 in utf-16 and 0x00 in utf-8. The UTF-8 string resulting from that conversion should include any NULs and any text after them, and if it's std::string, there's also an extra NUL past the end for null-termination).

Comment (jrprice) [reply src/webgpu/shader/validation/tokens.spec.ts:31]: Resolved offline - the tests are fine, there's an issue with our stack that we can address later.

File: src/webgpu/shader/validation/tokens.spec.ts [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+ ${t.params.terminated ? '*/' : ''}`;
+
+    t.expectCompileResult(t.params.terminated, code);
+  });
+
+const kValidIdentifiers = new Set(['foo', 'Foo', '_foo0', '_0foo', 'foo__0']);
+const kInvalidIdentifiers = new Set([
+  '_', // Single underscore is a syntactic token for phony assignment.
+  '__foo', // Leading double underscore is reserved.
+  '0foo', // Must start with single underscore or alpha-numeric.
```
Comment (zoddicus) [inline src/webgpu/shader/validation/tokens.spec.ts:95]: I am unclear why this invalid. `0foo` looks like it starts with an alpha-numeric character, `0`

Comment (jrprice) [reply src/webgpu/shader/validation/tokens.spec.ts]: The comment was wrong - the first character has to be underscore or a letter, not alphanumeric. I've fixed the comment.

**Review by zoddicus — APPROVED** [review]

**Pull request merged**